File: backend/src/main/java/com/workflow/framework/security/jwt/JwtTokenProvider.java
package com.workflow.framework.security.jwt;

import com.workflow.common.constant.GlobalConstant;
import io.jsonwebtoken.*;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.util.Date;

@Component
public class JwtTokenProvider {
    private final SecretKey secretKey;
    private final long accessTokenExpire;
    private final long refreshTokenExpire;

    public JwtTokenProvider(
            @Value("${jwt.secret:YnJfRFdpbjQ5dU5hZzJ4c0VvN0s2bUx4R0R4aUZ3cG1WeU9yS3F6RXhDdVN4V0E=}") String secret,
            @Value("${jwt.access-token-expire:30}") long accessTokenExpire,
            @Value("${jwt.refresh-token-expire:10080}") long refreshTokenExpire) {
        this.secretKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(secret));
        this.accessTokenExpire = accessTokenExpire;
        this.refreshTokenExpire = refreshTokenExpire;
    }

    public String createAccessToken(Long userId, String username) {
        Date now = new Date();
        return Jwts.builder()
                .subject(userId.toString())
                .claim("username", username)
                .claim("type", GlobalConstant.ACCESS_TOKEN_KEY)
                .issuedAt(now)
                .expiration(new Date(now.getTime() + accessTokenExpire * 60 * 1000))
                .signWith(secretKey)
                .compact();
    }

    public String createRefreshToken(Long userId, String username) {
        Date now = new Date();
        return Jwts.builder()
                .subject(userId.toString())
                .claim("username", username)
                .claim("type", GlobalConstant.REFRESH_TOKEN_KEY)
                .issuedAt(now)
                .expiration(new Date(now.getTime() + refreshTokenExpire * 60 * 1000))
                .signWith(secretKey)
                .compact();
    }

    public boolean validateToken(String token) {
        try {
            Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token);
            return true;
        } catch (JwtException | IllegalArgumentException e) {
            return false;
        }
    }

    public Long getUserIdFromToken(String token) {
        Claims claims = Jwts.parser()
                .verifyWith(secretKey)
                .build()
                .parseSignedClaims(token)
                .getPayload();
        return Long.parseLong(claims.getSubject());
    }

    public String getUsernameFromToken(String token) {
        Claims claims = Jwts.parser()
                .verifyWith(secretKey)
                .build()
                .parseSignedClaims(token)
                .getPayload();
        return claims.get("username", String.class);
    }

    public String getTokenType(String token) {
        Claims claims = Jwts.parser()
                .verifyWith(secretKey)
                .build()
                .parseSignedClaims(token)
                .getPayload();
        return claims.get("type", String.class);
    }
}

File: backend-node/src/framework/security/jwt-token.provider.ts
import { sign, verify, decode, type JwtPayload } from 'jsonwebtoken'
import type { JwtEnv } from '../config/env'

/** 对齐 Java GlobalConstant.ACCESS_TOKEN_KEY。 */
export const ACCESS_TOKEN_TYPE = 'access_token'
/** 对齐 Java GlobalConstant.REFRESH_TOKEN_KEY。 */
export const REFRESH_TOKEN_TYPE = 'refresh_token'

/**
 * JWT 签发与校验，逐字对齐 Java com.workflow.framework.security.jwt.JwtTokenProvider。
 *
 * 关键对齐点：
 *  - 密钥是 base64 字符串，需解码为 Buffer 后作为 HMAC 密钥。默认密钥解码后 47 字节，
 *    落在 jjwt 的 HMAC 分级区间（>=32 且 <48 字节 → HS256），因此算法是 **HS256**。
 *  - claims 恰为 sub（userId 的字符串形式）、username、type、iat、exp。
 *  - access 30 分钟 / refresh 10080 分钟，由配置决定。
 */
export class JwtTokenProvider {
  private readonly secret: Buffer
  private readonly accessTokenExpireSeconds: number
  private readonly refreshTokenExpireSeconds: number

  constructor(env: JwtEnv) {
    this.secret = Buffer.from(env.secret, 'base64')
    this.accessTokenExpireSeconds = env.accessTokenExpireMinutes * 60
    this.refreshTokenExpireSeconds = env.refreshTokenExpireMinutes * 60
  }

  createAccessToken(userId: number, username: string): string {
    return sign({ username, type: ACCESS_TOKEN_TYPE }, this.secret, {
      algorithm: 'HS256',
      subject: String(userId),
      expiresIn: this.accessTokenExpireSeconds,
    })
  }

  createRefreshToken(userId: number, username: string): string {
    return sign({ username, type: REFRESH_TOKEN_TYPE }, this.secret, {
      algorithm: 'HS256',
      subject: String(userId),
      expiresIn: this.refreshTokenExpireSeconds,
    })
  }

  validateToken(token: string): boolean {
    try {
      verify(token, this.secret, { algorithms: ['HS256'] })
      return true
    } catch {
      return false
    }
  }

  getUserIdFromToken(token: string): number {
    return Number(this.payload(token).sub)
  }

  getUsernameFromToken(token: string): string {
    return this.payload(token).username as string
  }

  getTokenType(token: string): string {
    return this.payload(token).type as string
  }

  private payload(token: string): JwtPayload & { username?: string; type?: string } {
    const decoded = decode(token)
    if (decoded === null || typeof decoded === 'string') {
      throw new Error('非法的 JWT')
    }
    return decoded as JwtPayload & { username?: string; type?: string }
  }
}
